# Teori Statistics
Distribution

- Random Variable

Random Variable adalah satu titik data, dari sample data yang kita miliki. Satu titik data (variable) yang kita ambil secara acak (random).

- Discrete Random Variable
- Continuous Random Variable

Discrete adalah saat value dari random variable tersebut bersifat bulat. Setiap nilai terpisah dengan jelas dari nilai lainnya. Tidak ada nilai di "antara" dua nilai. --> bar chart

Continuous adalah saat value dari random variable bersifat berjalan terus tanpa pemisah yang pasti. Data yang mengalir seperti ini tidak bisa kita hitung dengan mudah, biasanya kita perlu buat "bin" untuk range data tertentu, baru kita hitung ada berapa titik data yang masuk ke setiap bin berbeda. --> histogram


Untuk menentukan berapa kemungkinan kita mendapat value random variable tertentu dari dataset kita, kita mau membuat grafik distribusi

jenis data berbeda akan membentuk jenis distribusi berbeda, akan fokus ke distribusi normal.

# Teknik Inferential Statistics
1. Confidence Interval
2. Hypothesis Testing



### Confidence Interval

membuat "tengah" datanya menggunakan range. Sekaligus memberi informasi penyebaran datanya.

dibandingkan dengan menyebutkan rata2 tinggi 15 anak adalah 150 cm, kita akan buat range nya.

Formula Confidence Interval :

- best_estimate +- margin_of_error
- dari (best_estimate - margin_of_error) sampai (best_estimate + margin_of_error)


misalnya best_estimate tinggi badan pria (dari sample data) adl rata2 175cm
, dengan margin of error di 5cm, maka confidence intervalnya adalah 175-5cm sampai 175+5cm ===> 170 sampai 180 cm

untuk confidence_interval *rata-rata*
- best_estimate = rata-rata
- margin_of_error = z * standard_error
- standard_error = standard_deviation / sqrt(jumlah_data)

kenapa ada jumlah data di kalkulasi ? jumlah data mempengaruhi akurasi data yang dimiliki.

semakin besar jumlah datanya, maka standard error semakin kecil. asumsi makin banyak data maka lebih akurat

z adalah representasi seberapa yakin kita terhadap interval ini ? 
semakin yakin, maka z semakin besar. karena semakin kita yakin, misalnya 99% maka range yg kita kasih harus semakin besar, include semua populasi yang mungkin ada

untuk confidence_interval *proporsi*
- best_estimate = proporsi
- margin_of_error = z * standard_error
- standard_error = sqrt(proporsi * (1-proporsi) / jumlah_data)


## Case Study - Heart Disease dataset
kita akan gunakan data tentang pasien sakit jantung dari https://raw.githubusercontent.com/stefkwan-h8/dataset/refs/heads/main/heart.csv  (data aslinya dari https://www.kaggle.com/ronitf/heart-disease-uci ) untuk menghitung confidence interval

berikut beberapa konteks arti data di kolom tertentu
- sex (gender, 1 = Pria, 0 = Wanita)
- chest pain type (tipe sakit di dada, ada 4 jenis)
- fasting blood sugar > 120 mg/dl (tingkat gula darah >= 120 mg/dl, 1 = ya, 0 = tidak)
- exercise induced angina (merasakan sakit dada saat olahraga? 1 = ya, 0 = tidak)
- target (0 = sehat; 1 = sakit jantung)

In [2]:
import pandas as pd
import numpy as np

In [3]:
#import data

df_heart = pd.read_csv("https://raw.githubusercontent.com/stefkwan-h8/dataset/refs/heads/main/heart.csv ")
df_heart.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

In [4]:
df_heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


misalnya buat statement "orang yg sakit jantung tingkat kolestrolnya ada di angka..."

kita cari confidence interval rata2 kolestrol untuk pasien sakit jantung

In [5]:
df_sakit = df_heart[df_heart['target']==1]
df_sakit

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
5      58    0   0       100   248    0        0      122      0      1.0   
10     71    0   0       112   149    0        1      125      0      1.6   
12     34    0   1       118   210    0        1      192      0      0.7   
15     34    0   1       118   210    0        1      192      0      0.7   
16     51    0   2       140   308    0        0      142      0      1.5   
...   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
1011   45    1   1       128   308    0        0      170      0      0.0   
1014   44    0   2       108   141    0        1      175      0      0.6   
1019   47    1   0       112   204    0        1      143      0      0.1   
1020   59    1   1       140   221    0        1      164      1      0.0   
1023   50    0   0       110   254    0        0      159      0      0.0   

      slope  ca  thal  target  
5         1   0     2       1  
10        1   0     2       1  
12        2   0     2       1  
15        2   0     2       1  
16        2   1     2       1  
...     ...  ..   ...     ...  
1011      2   0     2       1  
1014      1   0     2       1  
1019      2   0     2       1  
1020      2   0     2       1  
1023      2   0     2       1  

[526 rows x 14 columns]

In [6]:
df_sakit['chol'].mean()

240.97908745247148

In [7]:
best_estimate = df_sakit['chol'].mean()

z = 1.96 #asumsi tingkat keyakinan di 95%
standard_deviation = df_sakit['chol'].std()
jumlah_data = df_sakit['chol'].count()
standard_error = standard_deviation / np.sqrt(jumlah_data)

margin_of_error = z * standard_error

print([best_estimate - margin_of_error, best_estimate + margin_of_error])

[236.4488201044286, 245.50935480051436]


In [14]:
#latihan cari confidence interval rata-rata untuk age , penderita penyakit jantung
#cari dulu yg sakit jantung (target = 1)
#hitung datanya berapa
#hitung rata-rata umur nya berapa
#standard deviation
#masukkin ke formula

#"biasanya yg menderita penyakit jantung ada di umur ..."

jumlah_data2 = df_sakit['age'].count() #yg sakit jantung (target = 1)
best_estimate2 = df_sakit['age'].mean() #rata-rata umur nya berapa
standard_deviation2 = df_sakit['age'].std() #standard deviation


standard_error2 = standard_deviation2 / np.sqrt(jumlah_data2)

margin_of_error2 = z * standard_error2

print([best_estimate2 - margin_of_error2, best_estimate2 + margin_of_error2])


[51.585610650913665, 53.23187984338291]


rata2 tingkat kolestrol pasien penderita penyakit jantung ada di antara 236 - 246

## sekarang kita masuk ke CI untuk proporsi

bagaimana kalau kita mau membuat statement yang lebih ke arah "xx% dari orang yang tingkat kolesterol nya di atas 200, adalah penderita penyakit jantung"

- yang kolesterol diatas 200 ada brp orang
- dari yg kolestrol > 200, ada berapa menderita penyakit jantung
- proporsi penderita penyakit jantung
- confidence interval proporsi

"berapa banyak bagian yg masuk dalam kategori tertentu?"

In [23]:
df_chol200 = df_heart[df_heart['chol'] > 200] #org yg kolestrol > 200

df_chol200_sakit = df_chol200[df_chol200['target'] == 1] #kolestrol > 200 & sakit
df_chol200_sehat = df_chol200[df_chol200['target'] == 0] #kolestrol > 200 & sehat

n_chol200 = df_chol200.shape[0]
n_sakit = df_chol200_sakit.shape[0]
p_sakit = n_sakit / n_chol200



In [24]:
# confidence interval

be = p_sakit

z = 1.96
se = np.sqrt(p_sakit * (1-p_sakit) / n_chol200)
me = z * se

[be - me, be + me]

[0.4670316286168878, 0.534140704325668]

### latihan pendek
coba cari confidence interval untuk proporsi.

dari wanita yang umurnya diatas 60 (para lansia wanita), ada berapa banyak proporsi yang menderita penyakit jantung?

.. - .. % lansia wanita menderita penyakit jantung.

- ambil dulu data lansia wanita saja
- hitung berapa dari data lansia wanita yang sakit
- hitung proporsi sakit / semua lansia wanita
- confidence interval proporsi

8.50pm WIB kita bahas

In [33]:
k_60 = (df_heart['age'] > 60)
k_wanita = (df_heart['sex']  == 0)
df_lansia = df_heart[k_60 & k_wanita] #lansia
df_lansia_sakit = df_lansia[df_lansia['target'] == 1 ] # lansia sakit
df_lansia_sakit.shape


(70, 14)

In [34]:
p_lansia_sakit = df_lansia_sakit.shape[0] / df_lansia.shape[0]
n_lansia = df_lansia.shape[0]

In [35]:
be = p_lansia_sakit

z = 1.96
se = np.sqrt(p_lansia_sakit * (1-p_lansia_sakit) / n_lansia)
me = z * se

[be - me, be + me]

[0.524668710506969, 0.7034014649316276]

# Hypothesis testing

membuat statement

contoh:
- pasien sakit lutut kalau terapi 2X seminggu sembuhnya lebih lama (daripada terapi 3X seminggu ~8.2 minggu baru sembuh)
- null hypothesis H0 (statement yang dianggap benar)
- alternate hypothesis H1 (statement baru yang mau kita buat, yang bisa diterima kalau kita menunjukkan bahwa H0 tidak benar)
- H1 sembuhnya lebih lama, sembuhnya > 8.2 minggu
- H0 menyatakan kalau sembuhnya <= 8.2 minggu

masalahnya kita tidak mau langsung membandingkan angka rata-rata kesembuhan pasien yang terapi 2X seminggu dengan angka rata-rata kesembuhan pasien yang terapi 3X seminggu. kita juga mau:
- masukkan informasi jumlah data yang kita miliki. lebih banyak data dianggap lebih akurat, lebih sedikit dianggap tidak seakurat itu.
- memilih tingkat keyakinan terhadap statement yang kita buat ini, karena tidak ada statement yagn 100% benar. biasanya kita set ke 95% yakin. atau tingkat error di 5%. `alpha` di 0.05. alpha adalah angka representative error rate yang kita gunakan untuk hypothesis testing.
- selain membandingkan angka, kita bandingkan `p-value` ke alpha. cara cari p-value kita gunakan bantuan package python.

kalau p_value > alpha, kita bilang FAIL to REJECT H0   
kalau p_value <= alpha, kita bilang REJECT H0

kita tidak pernah membuat pernyataan H1 itu benar, kita cuma bilang kalau ada perbedaan yang signifikan antara angka yang baru dengan asumsi lama H0 yang pernah dibuat. H1 mungkin benar. H0 kok beda jauh signifikan dari hasil hitung kita.



In [36]:
import statsmodels.api as sm
import matplotlib.pyplot as plt

# Case Study
- 52% ortu percaya X
- saat ini semakin banyak ortu percaya X
- H1 : p > 0.52 #skg lebih banyak yg percaya X, 52%
- H0 : p <= 0.52 #kebalikan dari H1
- sekarang dr data baru 1018, 56% ortu percaya X

In [38]:
n = 1018
phat = 0.56 #sekarang
pnull = 0.52 #dulu
#apakah bisa bilang skg lebih banyak ortu yg percaya X ?
#apakah %4 ini di bilang cukup berbeda ?
# 

apakah perbedaan 4% ini bisa dibilang cukup berbeda?

kalau % yang percaya sekarang 52.1%, apakah bisa dibilang berbeda?

kalau jumlah orang yang di survey sekarang 100rb orang, apakah perbedaan 0.1% ini signifikan?

kalau dari data kita yang di survey adalah 1018 orang, apakah perbedaan 4% ini signifikan?

In [39]:
# untuk perbandingan proporsi, dari 1 populasi

sm.stats. proportions_ztest(phat*n, n , pnull, alternative='larger')

(2.571067795759113, 0.0050692738658605335)